bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Bookie.java

```diff
@@ -441,77 +335,122 @@ private void checkIfDirsOnSameDiskPartition(List<File> dirs) throws DiskPartitio
-                try {
-                    Cookie journalCookie = Cookie.readFromDirectory(journalDirectory);
-                    journalCookies.add(journalCookie);
-                    if (journalCookie.isBookieHostCreatedFromIp()) {
-                        conf.setUseHostNameAsBookieID(false);
-                    } else {
-                        conf.setUseHostNameAsBookieID(true);
-                    }
-                } catch (FileNotFoundException fnf) {
-                    newEnv = true;
-                    missedCookieDirs.add(journalDirectory);
-                }
-            }
-
+            // 1. retrieve the instance id
             String instanceId = rm.getClusterInstanceId();
+
+            // 2. build the master cookie from the configuration
             Cookie.Builder builder = Cookie.generateCookie(conf);
             if (null != instanceId) {
                 builder.setInstanceId(instanceId);
             }
             Cookie masterCookie = builder.build();
-            Versioned<Cookie> rmCookie = null;
-            try {
-                rmCookie = Cookie.readFromRegistrationManager(rm, conf);
-                // If allowStorageExpansion option is set, we should
-                // make sure that the new set of ledger/index dirs
-                // is a super set of the old; else, we fail the cookie check
-                masterCookie.verifyIsSuperSet(rmCookie.getValue());
-            } catch (CookieNotFoundException e) {
-                // can occur in cases:
-                // 1) new environment or
-                // 2) done only metadata format and started bookie server.
-            }
-            for (File journalDirectory : journalDirectories) {
-                checkDirectoryStructure(journalDirectory);
-            }
-            if (!newEnv) {
-                for (Cookie journalCookie: journalCookies) {
-                    masterCookie.verifyIsSuperSet(journalCookie);
-                }
-            }
 
+            // 3. read the cookie from registration manager. it is the `source-of-truth` of a given bookie.
+            //    if it doesn't exist in registration manager, this bookie is a new bookie, otherwise it is
+            //    an old bookie.
+            final Versioned<Cookie> rmCookie = readAndVerifyCookieFromRegistrationManager(
+                masterCookie,
+                conf,
+                rm
+            );
+            final boolean newEnv = null == rmCookie;
 
-            for (File dir : allLedgerDirs) {
-                checkDirectoryStructure(dir);
-                try {
-                    Cookie c = Cookie.readFromDirectory(dir);
-                    masterCookie.verifyIsSuperSet(c);
-                } catch (FileNotFoundException fnf) {
-                    missedCookieDirs.add(dir);
-                }
+            // 4. check if the cookie appear in all the directories.
+            List<File> missedCookieDirs = new ArrayList<>();
```

Comment:
I think this algorithm would be much simpler if moved the special cases for newEnv and expandStorage somewhere else.

In the case of new env, just add cookies to all the dirs.

In the case of expand storage:
- check that all the dirs in rmCookie have a cookie equal to rmCookie
- all the dirs in masterCookie that are not in rmCookie are empty
- Update the cookie in all dirs and zk

Then in all cases run the verification that all dirs contain the same cookie as zk (in new and expand cases, they will have been updated).

Reply:
please see my comment above regarding newEnv. I would prefer not simplifying the validation for newEnv. I removed the branches is to ensure validation in the common path. because if you do this branch, it is hard to enforce changes in future.